atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/ConflictDetectionManager.java

```diff
@@ -15,56 +15,133 @@
  */
 package com.palantir.atlasdb.transaction.impl;
 
-import java.util.Collections;
-import java.util.HashMap;
 import java.util.Map;
-import java.util.function.Supplier;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
 import com.google.common.collect.Maps;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.table.description.TableMetadata;
 import com.palantir.atlasdb.transaction.api.ConflictHandler;
 
 public class ConflictDetectionManager {
-    private final Supplier<Map<TableReference, ConflictHandler>> supplier;
-    private final Map<TableReference, ConflictHandler> overrides;
-    private final RecomputingSupplier<Map<TableReference, ConflictHandler>> cachedResult;
+    private static final Logger log = LoggerFactory.getLogger(ConflictDetectionManager.class);
+    private final LoadingCache<TableReference, ConflictHandler> cache;
 
     /**
-     * @param supplier A {@link Supplier} which returns the current conflict detection
-     *                 configuration to the database. Note that the result is cached,
-     *                 so changes to configuration must be followed by a call to
-     *                 {@link #recompute()} in order to be picked up.
+     *  This class does not make the mistake of attempting cache invalidation,
+     *  so a table dropped by another instance may still be cached here.
+     *
+     *  This is okay in the case of a simple drop, but a same-name table drop
+     *  and re-addition with a different Conflict Handler
+     *  (where an external atlas instance handles both of these operations)
+     *  will be incorrect. This is an unrealistic workflow
+     *  and I'm fine with just documenting this behavior.
+     *
+     *  (This has always been the behavior of this class; I'm simply calling it out)
      */
-    ConflictDetectionManager(Supplier<Map<TableReference, ConflictHandler>> supplier) {
-        this.supplier = supplier;
-        this.overrides = Maps.newConcurrentMap();
-        this.cachedResult = RecomputingSupplier.create(() -> {
-            Map<TableReference, ConflictHandler> ret = new HashMap<>(this.supplier.get());
-            ret.putAll(overrides);
-            return Collections.unmodifiableMap(ret);
-        });
+    private ConflictDetectionManager(CacheLoader<TableReference, ConflictHandler> loader) {
+        this.cache = CacheBuilder.newBuilder()
+                .maximumSize(100_000)
```

Comment:
Maybe we could add an expiration time here to make a bit less wrong in case the drop-recreate scenario actually happens